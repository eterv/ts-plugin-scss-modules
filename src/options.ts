import { Options as SassOptions } from 'sass';
import tsModule from 'typescript/lib/tsserverlibrary';
import { DotenvConfigOptions } from 'dotenv/types';
import { CSSExports } from 'icss-utils';
import { Logger } from './helpers/logger';

export interface PostcssOptions {
  excludePlugins?: string[];
  useConfig?: boolean;
}

export interface RendererOptions {
  sass?: Partial<SassOptions<'sync'>>;
}

export interface Options {
  classnameTransform?: ClassnameTransformOptions;
  customMatcher?: string;
  customRenderer?: string;
  customTemplate?: string;
  dotenvOptions?: DotenvConfigOptions;
  goToDefinition?: boolean;
  namedExports?: boolean;
  postcssOptions?: PostcssOptions;
  /** @deprecated To align with other projects. */
  postCssOptions?: PostcssOptions;
  rendererOptions?: RendererOptions;
}

export type ClassnameTransformOptions =
  | 'asIs'
  | 'camelCase'
  | 'camelCaseOnly'
  | 'dashes'
  | 'dashesOnly';

export interface CustomRendererOptions {
  fileName: string;
  logger: Logger;
  compilerOptions: tsModule.CompilerOptions;
}

export type CustomRenderer = (
  css: string,
  options: CustomRendererOptions,
) => string;

export interface CustomTemplateOptions {
  classes: CSSExports;
  fileName: string;
  logger: Logger;
}

export type CustomTemplate = (
  dts: string,
  options: CustomTemplateOptions,
) => string;
