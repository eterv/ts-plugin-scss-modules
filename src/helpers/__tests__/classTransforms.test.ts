import { transformClasses } from '../classTransforms';
import { ClassnameTransformOptions } from '../../options';

describe('utils / classTransforms', () => {
  const classNames = [
    'class-name-a',
    'classNameB',
    'class-Name-C',
    '__class_nAmeD--',
  ];
  const tests: ClassnameTransformOptions[] = [
    'camelCase',
    'camelCaseOnly',
    'dashes',
    'dashesOnly',
    'asIs',
  ];

  it(`should not transform classes when no option is set`, () => {
    const transformer = transformClasses();
    const transformedClasses = classNames.map(transformer);
    expect(transformedClasses).toMatchSnapshot();
  });

  tests.forEach((option) => {
    it(`should transform classes correctly when \`classnameTransform\` set to \`${option}\``, () => {
      const transformer = transformClasses(option);
      const transformedClasses = classNames.map(transformer);
      expect(transformedClasses).toMatchSnapshot();
    });
  });
});
