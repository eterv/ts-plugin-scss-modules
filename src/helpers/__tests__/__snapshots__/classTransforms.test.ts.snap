// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils / classTransforms should not transform classes when no option is set 1`] = `
Array [
  Array [
    "class-name-a",
  ],
  Array [
    "classNameB",
  ],
  Array [
    "class-Name-C",
  ],
  Array [
    "__class_nAmeD--",
  ],
]
`;

exports[`utils / classTransforms should transform classes correctly when \`classnameTransform\` set to \`asIs\` 1`] = `
Array [
  Array [
    "class-name-a",
  ],
  Array [
    "classNameB",
  ],
  Array [
    "class-Name-C",
  ],
  Array [
    "__class_nAmeD--",
  ],
]
`;

exports[`utils / classTransforms should transform classes correctly when \`classnameTransform\` set to \`camelCase\` 1`] = `
Array [
  Array [
    "class-name-a",
    "classNameA",
  ],
  Array [
    "classNameB",
  ],
  Array [
    "class-Name-C",
    "classNameC",
  ],
  Array [
    "__class_nAmeD--",
    "classNAmeD",
  ],
]
`;

exports[`utils / classTransforms should transform classes correctly when \`classnameTransform\` set to \`camelCaseOnly\` 1`] = `
Array [
  Array [
    "classNameA",
  ],
  Array [
    "classNameB",
  ],
  Array [
    "classNameC",
  ],
  Array [
    "classNAmeD",
  ],
]
`;

exports[`utils / classTransforms should transform classes correctly when \`classnameTransform\` set to \`dashes\` 1`] = `
Array [
  Array [
    "class-name-a",
    "classNameA",
  ],
  Array [
    "classNameB",
  ],
  Array [
    "class-Name-C",
    "classNameC",
  ],
  Array [
    "__class_nAmeD--",
  ],
]
`;

exports[`utils / classTransforms should transform classes correctly when \`classnameTransform\` set to \`dashesOnly\` 1`] = `
Array [
  Array [
    "classNameA",
  ],
  Array [
    "classNameB",
  ],
  Array [
    "classNameC",
  ],
  Array [
    "__class_nAmeD--",
  ],
]
`;
